std-bits/base/src/main/java/org/enso/base/Text_Utils.java

```diff
@@ -162,29 +158,83 @@ public static int[] get_codepoints(String str) {
    * @return array of substrings of {@code str} contained between occurences of {@code sep}
    */
   public static String[] split_by_literal(String str, String sep) {
-    return str.split(Pattern.quote(sep));
+    return str.split(Pattern.quote(sep), -1);
   }
 
   /**
    * Splits the string on each occurrence of UTF-8 whitespace, returning the resulting substrings in
    * an array.
    *
    * @param str the string to split
+   * @param keep_whitespace whether to keep the whitespace as separate entries
    * @return the array of substrings of {@code str}
    */
-  public static String[] split_on_whitespace(String str) {
-    return whitespace.split(str);
+  public static List<String> split_on_whitespace(String str, boolean keep_whitespace) {
+    ArrayList<String> acc = new ArrayList<>();
+    int length = str.length();
+    int currentStart = 0;
+    boolean isAccumulatingWhitespace = UCharacter.isWhitespace(str.charAt(0));
+    int currentPos = 1;
+    while (currentPos < length) {
+      boolean encounteredWhitespace = UCharacter.isWhitespace(str.charAt(currentPos));
+      if (encounteredWhitespace != isAccumulatingWhitespace) {
+        if (currentPos > currentStart) {
+          if (!isAccumulatingWhitespace || keep_whitespace) {
+            acc.add(str.substring(currentStart, currentPos));
+          }
+        }
+
+        currentStart = currentPos;
+        isAccumulatingWhitespace = encounteredWhitespace;
+      }
+
+      currentPos++;
+    }
+
+    if (!isAccumulatingWhitespace || keep_whitespace) {
+      acc.add(str.substring(currentStart));
+    }
+
+    return acc;
   }
 
   /**
    * Splits the string on each occurrence of UTF-8 vertical whitespace, returning the resulting
    * substrings in an array.
    *
    * @param str the string to split
+   * @param keep_endings whether to keep line endings in returned lines
    * @return the array of substrings of {@code str}
    */
-  public static String[] split_on_lines(String str) {
-    return vertical_space.split(str);
+  public static List<String> split_on_lines(String str, boolean keep_endings) {
+    ArrayList<String> acc = new ArrayList<>();
+    int length = str.length();
+    int currentStart = 0;
+    int currentPos = 0;
+    while (currentPos < length) {
```

Comment:
Why not use the String.indexOf method for this loop?

Reply:
It was just conceptually simpler for me to think of this as a looping through every character (`indexOf` will do it anyway so performance-wise its the same). But I'm not very tied to this solution so I can changed to indexOf if you think that's better.

Reply:
seemed no reason to reproduce the logic of the indexOf function - as not gaining anything here

Reply:
But it's not really reproducing its logic - it is a similar traversal but more is happening here. Especially with handling the `\r\n` 'digraph'. I can write an alternative approach and let's compare it.